python/vyos/firewall.py

```diff
@@ -366,6 +376,147 @@ def parse_policy_set(set_conf, def_suffix):
+                reader = csv.reader(csv_fh)
+                for entry in reader:
+                    out.append([list, entry[0]])
+        except Exception as e:
+            print(f'Error: Failed to open {list} list database.\n{e}')
+            return []
+    return out
+
+def list_download(list=None, url=None):
+    if not list or not url:
+        return False
+
+    try:
+        if not os.path.exists(lists_database_dir):
+            os.mkdir(lists_database_dir)
+
+        list_database = os.path.join(lists_database_dir, list + '.csv')
+        if url.startswith(os.sep):
+            if not os.path.exists(url):
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), url)
+            shutil.copyfile(url, list_database)
+        else:
+          download(list_database, url)
+        print(f'Downloaded {list} list database')
+        return True
+    except Exception as e:
+        print(f'Error: Failed to download {list} list database.\n{e}')
+    return False
+
+class ListsLock(object):
+    def __init__(self, file):
+        self.file = file
+
+    def __enter__(self):
+        if os.path.exists(self.file):
+            return False
+
+        Path(self.file).touch()
+        return True
+
+    def __exit__(self, exc_type, exc_value, tb):
+        os.unlink(self.file)
+
+def list_update(firewall, name=None, force=False):
+    with ListsLock(lists_lock_file) as lock:
+        if not lock:
+            print('Script is already running')
+            return False
+
+        if not firewall:
+            print('Firewall is not configured')
+            return True
+
+        if 'lists' not in firewall['group']:
+            print('No lists are configured')
+            return True
+
+        names = []
+        tasks = []
+        for key in firewall['group']['lists']:
+            interval = None if 'interval' not in firewall['group']['lists'][key] else firewall['group']['lists'][key]['interval']
+            spec = None if 'crontab-spec' not in firewall['group']['lists'][key] else firewall['group']['lists'][key]['crontab-spec']
+            if interval or spec:
+                task = {
+                        "name": key,
+                        "interval": interval,
+                        "spec": spec,
+                        "executable": '/usr/libexec/vyos/list-update.py',
+                        "args": f'--force --name {key}'
+                       }
+                tasks.append(task)
+            if not name or name == key:
+                if not os.path.exists(os.path.join(lists_database_dir, key + '.csv')) or force:
+                    if not list_download(key, firewall['group']['lists'][key]['url']):
+                        continue
+                names.append(key)
+
+        try:
+            # Maybe task_scheduler logic should be moved into vyos package?
+            import sys
+            sys.path.append('/usr/libexec/vyos/conf_mode')
+            import task_scheduler
```

Comment:
Yes I made mention of this in IRC, it's not very good practice, but since this is my first contribution, I didn't want to refactor all of task-scheduler to move it into the vyos python package.  But I agree it should be done as my comment states.  Didn't want my changes to get lost by combining issues.

That being said I'd be happy to make a sub task to move it, then I could `from vyos.task-scheduler import verify, generate, apply` or maybe even make a convince function to do all that in one call.

Reply:
I think it's best to move just the cron format code to the vyos module and reference that in both this script and in task scheduler, though perhaps that can done after this initial PR.

Reply:
Moved the code so this could be imported and used properly